ebms-async/src/main/kotlin/no/nav/emottak/ebms/async/kafka/consumer/FailedMessageKafkaHandler.kt

```
@@ -0,0 +1,220 @@
+import org.apache.kafka.common.serialization.StringDeserializer
+import org.apache.kafka.common.serialization.StringSerializer
+import org.joda.time.DateTime
+import org.slf4j.LoggerFactory
+import java.math.BigInteger
+import java.time.Duration
+import kotlin.time.Duration.Companion.seconds
+
+val failedMessageQueue: FailedMessageKafkaHandler = FailedMessageKafkaHandler()
+const val RETRY_COUNT_HEADER = "retryCount"
+const val RETRY_AFTER = "retryableAfter"
+const val RETRY_REASON = "retryReason"
+
+val logger = LoggerFactory.getLogger(FailedMessageKafkaHandler::class.java)
+
+class FailedMessageKafkaHandler(
+    val kafkaErrorQueue: KafkaErrorQueue = config().kafkaErrorQueue,
+    kafka: Kafka = config().kafka
+) {
+
+    val publisher = KafkaPublisher(
+        PublisherSettings(
+            bootstrapServers = kafka.bootstrapServers,
+            keySerializer = StringSerializer(),
+            valueSerializer = ByteArraySerializer(),
+            acknowledgments = Acks.All,
+            properties = kafka.toProperties()
+        )
+    )
+
+    private var errorTopicKafkaReceiver = KafkaReceiver(
+        ReceiverSettings(
+            bootstrapServers = kafka.bootstrapServers,
+            keyDeserializer = StringDeserializer(),
+            valueDeserializer = ByteArrayDeserializer(),
+            groupId = kafka.groupId,
+            pollTimeout = 10.seconds,
+            properties = kafka.toProperties()
+        )
+    )
+
+    suspend fun sendToRetry(
+        record: ReceiverRecord<String, ByteArray>,
+        key: String = record.key(),
+        value: ByteArray = record.value(),
+        reason: String? = null
+    ) {
+        if (reason != null) {
+            record.addHeader(RETRY_REASON, reason)
+        }
+        record.addHeader(RETRY_AFTER, getNextRetryTime(record))
+        try {
+            val metadata = publisher.publishScope {
+                publish(ProducerRecord(config().kafkaErrorQueue.topic, null, key, value, record.headers()))
```

Hvis du heller bruker `publishCatching` - funksjonen her får du tilbake et `Result<Metadata>` så slipper du denne catch all handleren som bryter hele flyten.